File: Python3/543. Diameter_of_Binary_Tree.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None


class Solution:
    def __init__(self):
        self.ans = 0

    def diameterOfBinaryTree(self, root: TreeNode) -> int:
        if root == None:
            return 0
        self.depthOfTree(root)
        return self.ans

    def depthOfTree(self, root: TreeNode) -> int:
        if root == None:
            return 0
        left = self.depthOfTree(root.left)
        right = self.depthOfTree(root.right)
        ans = max(ans, left + right)
        return 1 + max(left, right)

File: Java/543.Diameter_of_Binary_Tree.java
/**
 * Definition for a binary tree node. public class TreeNode { int val; TreeNode
 * left; TreeNode right; TreeNode(int x) { val = x; } }
 */
class Solution {
    private int ans = 0;

    public int diameterOfBinaryTree(TreeNode root) {
        depthOfBinaryTree(root);
        return ans;
    }

    private int depthOfBinaryTree(TreeNode root) {
        if (root == null)
            return 0;
        int left = depthOfBinaryTree(root.left);
        int right = depthOfBinaryTree(root.right);
        ans = Math.max(ans, left + right);

        return 1 + Math.max(left, right);
    }
}